### PR TITLE
[core][flink] Introduce orphan files clean and corresponding flink action

### LINKED ISSUE #1571: [Feature] Introduce file clean action
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

Currently, the `FileStoreExpireImpl` delete file quietly, which means if there occurs error, the process won't throw exceptions. It's possible that some files won't be deleted when snapshot expiring, so I propose to introduce an action to clean files unused by current snapshots. 

### Solution

Scan all files and filter out that still used by at least one snapshot, then delete those not used by any snapshot.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### LINKED ISSUE #2222: [Bug]：Cleaning up residual orc files
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Paimon version

0.6-SNAPSHOT

### Compute Engine

flink 1.17

### Minimal reproduce step

1. paimon-flink-action executes mysql-sync-table
2. Writing to the orc file is completed and the job is canceled before the meta information is written
orc file becomes an orphan and cannot be deleted

### What doesn't meet your expectations?

Orphan orc files can be cleaned up, or a snapshot of the currently canceled job completed

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by yuzelin**
### Purpose

Linked issue: close #1571 

Fix #2222

### Tests

`OrphanFilesCleanTest`
`RemoveOrphanFilesActionITCase`

### Documentation
Understand Files - Remove Orphan Files

### CODE REVIEW

**Comment by yuzelin** [comment]
Some codes have referenced #2223

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+
+        DataStream<Row> allFilesDataStream =
+                env.fromCollection(
+                                listPaimonFileDirs(
+                                        table.fileIO(),
+                                        table.location(),
+                                        table.partitionKeys().size()))
+                        .flatMap(new ListCandidates(table.fileIO(), olderThanMillis))
+                        .returns(rowTypeInfo);
+        batchTEnv.createTemporaryView("ALL", allFilesDataStream);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java:374]: Can we just use `COMMON_IO_FORK_JOIN_POOL`? We don't need to use Flink Datastream

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+                "SELECT path FROM `ALL` WHERE `ALL`.fileName NOT IN (SELECT fileName FROM `USED`)";
+        CloseableIterator<Row> iterator = batchTEnv.executeSql(query).collect();
+
+        while (iterator.hasNext()) {
+            String fileName = iterator.next().getFieldAs(0);
+            fileStoreTable.fileIO().deleteQuietly(new Path(fileName));
+            deletedFileNums++;
+        }
+
+        iterator.close();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java:149]: Delete snapshot temporary files too.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+            latestSnapshotId =
+                    Math.max(latestSnapshotId, taggedSnapshots.get(tags.size() - 1).id());
+        }
+
+        FileStore<?> store = table.store();
+        ManifestList manifestList = store.manifestListFactory().create();
+        ManifestFile manifestFile = store.manifestFileFactory().create();
+        IndexFileHandler indexFileHandler = store.newIndexFileHandler();
+        Set<String> usedFiles = new HashSet<>();
+        for (long id = latestSnapshotId; id >= earliestSnapshotId; id--) {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java:175]: just construct a Set<Snapshot> to contain snapshots and tags.

File: paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java [hunk 4]
```diff
@@ -236,4 +238,14 @@ private static int findPreviousOrEqualTag(
         }
         return -1;
     }
+
+    @Nullable
+    public static Snapshot find(SortedMap<Snapshot, String> tags, long targetSnapshotId) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java:243]: remove this method.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+     * List all the used files by snapshots and tags, and store them in view USED(fileName STRING).
+     */
+    private void buildUsedFilesView(FileStoreTable table) throws IOException {
+        SnapshotManager snapshotManager = table.snapshotManager();
+        TagManager tagManager = table.tagManager();
+
+        long earliestSnapshotId = snapshotManager.earliestSnapshotId();
+        long latestSnapshotId = snapshotManager.latestSnapshotId();
+
+        SortedMap<Snapshot, String> tags = tagManager.tags();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java:162]: tags should ignore file not found exception

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+ *
+ * <p>To avoid conflicting with snapshot expiration, tag deletion and rollback, it will skip the
+ * snapshot/tag that are not existed when listing used files.
+ *
+ * <p>To avoid deleting files that are used but not read by mistaken, it will stop removing process
+ * when failed to read used files.
+ *
+ * <p>Some ideas refer to Iceberg's {@code DeleteOrphanFilesSparkAction}.
+ */
+public class RemoveOrphanFilesAction extends TableActionBase {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java:83]: move this to paimon-core

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java (hunk 3 above)
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java:175]: use `COMMON_IO_FORK_JOIN_POOL` to accelate.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+                caught = origin;
+            }
+            try {
+                TimeUnit.MILLISECONDS.sleep(READ_FILE_RETRY_INTERVAL);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        }
+        return ReadingResult.forException(caught);
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/action/RemoveOrphanFilesAction.java:299]: just skip FileNot found exception

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**